sys/sysdown/src/sysdown.c

```diff
@@ -0,0 +1,169 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "os/mynewt.h"
+
+/**
+ * The SYSDOWN setting is injected by new-enough versions of the newt tool.
+ */
+#if !MYNEWT_VAL(SYSDOWN)
+
+int
+sysdown(int reason, sysdown_complete_fn *complete_cb, void *arg)
+{
+    return 0;
+}
+
+#else
+
+#include <assert.h>
+#include <stdio.h>
+#include <stddef.h>
+#include <limits.h>
+#include "hal/hal_system.h"
+#include "hal/hal_watchdog.h"
+
+#define SYSDOWN_TIMEOUT_TICKS   \
+    (MYNEWT_VAL(SYSDOWN_TIMEOUT_MS) * OS_TICKS_PER_SEC / 1000)
+
+_Static_assert(SYSDOWN_TIMEOUT_TICKS >= 0 && SYSDOWN_TIMEOUT_TICKS < INT32_MAX,
+               "SYSDOWN_TIMEOUT_MS value not in valid range");
+
+sysdown_complete_fn *sysdown_complete_cb;
+void *sysdown_complete_arg;
+
+static volatile int sysdown_num_in_progress;
+bool sysdown_active;
+static struct os_callout sysdown_timer;
+
+static void
+sysdown_dflt_panic_cb(const char *file, int line, const char *func,
+                      const char *expr, const char *msg)
+{
+#if MYNEWT_VAL(SYSDOWN_PANIC_MESSAGE)
+    if (msg != NULL) {
+        fprintf(stderr, "sysdown failure: %s\n", msg);
+    }
+#endif
+
+    __assert_func(file, line, func, expr);
+}
+
+sysdown_panic_fn *sysdown_panic_cb = sysdown_dflt_panic_cb;
+
+/**
+ * Sets the sysdown panic function; i.e., the function which executes when
+ * shutdown fails.  By default, a panic triggers a failed assertion.
+ */
+void
+sysdown_panic_set(sysdown_panic_fn *panic_cb)
+{
+    sysdown_panic_cb = panic_cb;
+}
+
+static void
+sysdown_complete(void)
+{
+    os_callout_stop(&sysdown_timer);
+
+    if (sysdown_complete_cb != NULL) {
+        sysdown_complete_cb(0, sysdown_complete_arg);
+    }
+}
+
+void
+sysdown_release(void)
+{
+    os_sr_t sr;
+    int count;
+
+    OS_ENTER_CRITICAL(sr);
+    count = --sysdown_num_in_progress;
+    OS_EXIT_CRITICAL(sr);
+
+    if (count <= 0) {
+        sysdown_complete();
+    }
+}
+
+static void
+sysdown_timer_exp(struct os_event *unused)
+{
+    assert(0);
+}
+
+int
+sysdown(int reason, sysdown_complete_fn *complete_cb, void *arg)
+{
+    os_sr_t sr;
+    int rc;
+    int i;
+
+    /* Only allow one shutdown operation. */
+    OS_ENTER_CRITICAL(sr);
+    if (sysdown_active) {
+        rc = SYS_EALREADY;
+    } else {
+        sysdown_active = true;
+        rc = 0;
+    }
+    OS_EXIT_CRITICAL(sr);
+
+    if (rc != 0) {
+        return rc;
+    }
+
+    sysdown_complete_cb = complete_cb;
+    sysdown_complete_arg = arg;
+
+    os_callout_init(&sysdown_timer, os_eventq_dflt_get(), sysdown_timer_exp,
+                    NULL);
+    rc = os_callout_reset(&sysdown_timer, SYSDOWN_TIMEOUT_TICKS);
+    assert(rc == 0);
+
+    /* Call each configured sysdown callback. */
+    for (i = 0; sysdown_cbs[i] != NULL; i++) {
```

Comment:
Could it be sizeof(sysdown_cbs) / sizeof(sysdown_cbs[0])? Would save 4 bytes of text, if sysdown_cbs[] we're not terminated by a NULL...

Reply:
The reason I null-terminated the array is to handle the case where there aren't any sysdown procedures (a zero-length array being undefined).  gcc does allow zero-length arrays, but I didn't think four bytes' savings justified using an extension.

Reply:
That's fair. Agreed.